modules/debounce.js

```diff
@@ -1,35 +1,41 @@
 import restArguments from './restArguments.js';
-import delay from './delay.js';
+import now from './now.js';
 
 // When a sequence of calls of the returned function ends, the argument
 // function is triggered. The end of a sequence is defined by the `wait`
 // parameter. If `immediate` is passed, the argument function will be
 // triggered at the beginning of the sequence instead of at the end.
 export default function debounce(func, wait, immediate) {
-  var timeout, result;
+  var timeout, previous, args, result, context;
 
-  var later = function(context, args) {
-    timeout = null;
-    if (args) result = func.apply(context, args);
-  };
+  var later = function() {
+    var passed = now() - previous;
+    if (wait >= passed) {
+      timeout = setTimeout(later, wait - passed);
 
-  var debounced = restArguments(function(args) {
-    if (timeout) clearTimeout(timeout);
-    if (immediate) {
-      var callNow = !timeout;
-      timeout = setTimeout(later, wait);
-      if (callNow) result = func.apply(this, args);
     } else {
-      timeout = delay(later, wait, this, args);
+      timeout = null;
+      if (!immediate) result = func.apply(context, args);
+      // This check is needed because the argument function can recursively invoke debounced
+      if (!timeout) args = context = null;
     }
+  };
 
+  var debounced = restArguments(function(_args) {
+    context = this;
+    args = _args;
+    previous = now();
+    if (!timeout) {
+      timeout = setTimeout(later, wait);
+      if (immediate) result = func.apply(context, args);
+    }
     return result;
   });
 
   debounced.cancel = function() {
     clearTimeout(timeout);
-    timeout = null;
+    timeout = args = context = null;
   };
 
   return debounced;
-}
+}
```

Comment:
Please put back the newline at the end of the file. I know editors are nowadays often configured to strip them, but it is terminal-friendlier and concatenation-friendlier to always have a newline at the very end of a text file.

Reply:
Sorry, I thought it was like that original. I prefer newline at the end of the file.

Reply:
No problem!